core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.Constants;
+import alluxio.client.BoundedStream;
+import alluxio.client.Cancelable;
+import alluxio.client.block.BlockWorkerClient;
+import alluxio.client.file.FileSystemContext;
+import alluxio.client.file.options.OutStreamOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.proto.dataserver.Protocol;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.io.Closer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FilterOutputStream;
+import java.io.IOException;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Provides a stream API to write a block to Alluxio. An instance of this class can be obtained by
+ * calling
+ * {@link alluxio.client.block.AlluxioBlockStore#getOutStream(long, long, OutStreamOptions)}.
+ */
+@NotThreadSafe
+public final class BlockOutStream extends FilterOutputStream implements BoundedStream, Cancelable {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final long mBlockId;
+  private final long mBlockSize;
+  private final Closer mCloser;
+  private final BlockWorkerClient mBlockWorkerClient;
+  private final PacketOutStream mOutStream;
+  private boolean mClosed;
+
+  /**
+   * Creates a new local block output stream.
+   *
+   * @param blockId the block id
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the file system context
+   * @param options the options
+   * @throws IOException if an I/O error occurs
+   * @return the {@link BlockOutStream} instance created
+   */
+  public static BlockOutStream createLocalBlockOutStream(long blockId,
+      long blockSize,
+      WorkerNetAddress workerNetAddress,
+      FileSystemContext context,
+      OutStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
```

Comment:
can you try-with-resource and put the closer into the resource clause? if so, apply it throughout this PR